torchtune/modules/vq_embeddings.py

```diff
@@ -0,0 +1,173 @@
+        ).mul_(n)
+        # Get all encoded vectors attracted to each embedding vector
+        encoded_per_codebook = torch.matmul(codebook_onehot.t(), z)
+        # Update each embedding vector with new encoded vectors that are attracted to it,
+        # divided by its usage to yield the mean of encoded vectors that choose it
+        self.code_avg.mul_(self._decay).add_(
+            encoded_per_codebook, alpha=(1 - self._decay)
+        )
+        self.embedding = self.code_avg / self.code_usage.unsqueeze(1)
+        # Reset any embedding vectors that fall below threshold usage with random encoded vectors
+        z_rand = self._get_random_vectors(z, self.num_embeddings)
+        self.embedding = torch.where(
+            self.code_usage.unsqueeze(1) >= self.codebook_usage_threshold,
+            self.embedding,
+            z_rand,
+        )
+
+    def forward(self, z: Tensor) -> Tuple[Tensor, Tensor]:
+        """
+        Args:
+            z (Tensor): Tensor containing a batch of encoder outputs of shape ``(b, s, d)``, where
+                b is batch size, s is sequence length or time, and d is ``embedding_dim``.
+
+        Returns:
+            Tuple[Tensor, Tensor]: The quantized input and the embedding vector ids that were used.
+
+        Raises:
+            ValueError: if input embedding dimension does not match embedding dimension of module
+        """
+        bsz, seq_len, z_embed_dim = z.shape
+        if z_embed_dim != self.embedding_dim:
+            raise ValueError(
+                f"Expected last dimension of input tensor ({z_embed_dim}) to be embedding size of {self.embedding_dim}"
+            )
+
+        # Flatten into batch dimension
+        z_flat = z.view(-1, z_embed_dim)
+        # Calculate distances from each encoder, E(x), output vector to each embedding vector, e, ||E(x) - e||^2
+        distances = torch.cdist(z_flat, self.embedding, p=2.0) ** 2
+
+        # Encoding - select closest embedding vectors, shape [b * s, ]
+        token_ids_flat = torch.argmin(distances, dim=1)
+
+        # Quantize - shape [b * s, d]
+        quantized_flat = self.lookup(token_ids_flat)
+
+        # Use exponential moving average to update the embedding instead of a codebook loss,
+        # as suggested by Oord et al. 2017 and Razavi et al. 2019.
+        if self.training and self.learnable:
+            self._ema_update_embedding(z_flat, token_ids_flat)
+
+        # Straight through estimator
+        quantized_flat = z_flat + (quantized_flat - z_flat).detach()
+
+        # Reshape to original - [b, s, d] and [b, s]
+        quantized = quantized_flat.view(bsz, seq_len, z_embed_dim)
+        token_ids = token_ids_flat.view(bsz, seq_len)
+
+        return quantized, token_ids
+
+    def extra_repr(self) -> str:
+        return "num_embeddings={}, embedding_dim={}".format(
+            self.num_embeddings, self.embedding_dim
+        )
+
+    def lookup(self, token_ids: Tensor) -> Tensor:
```

Comment:
nit: do we want to use "lookup" here or something more consistent with the rest of the library like deocde?